drivers/serial/uart_lpc11u6x.c

```diff
@@ -0,0 +1,996 @@
+	.fifo_fill = lpc11u6x_uartx_fifo_fill,
+	.fifo_read = lpc11u6x_uartx_fifo_read,
+	.irq_tx_enable = lpc11u6x_uartx_irq_tx_enable,
+	.irq_tx_disable = lpc11u6x_uartx_irq_tx_disable,
+	.irq_tx_ready = lpc11u6x_uartx_irq_tx_ready,
+	.irq_tx_complete = lpc11u6x_uartx_irq_tx_complete,
+	.irq_rx_enable = lpc11u6x_uartx_irq_rx_enable,
+	.irq_rx_disable = lpc11u6x_uartx_irq_rx_disable,
+	.irq_rx_ready = lpc11u6x_uartx_irq_rx_ready,
+	.irq_err_enable = lpc11u6x_uartx_irq_err_enable,
+	.irq_err_disable = lpc11u6x_uartx_irq_err_disable,
+	.irq_is_pending = lpc11u6x_uartx_irq_is_pending,
+	.irq_update = lpc11u6x_uartx_irq_update,
+	.irq_callback_set = lpc11u6x_uartx_irq_callback_set,
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+
+#define LPC11U6X_UARTX_INIT(idx)                                              \
+									      \
+static const struct lpc11u6x_uartx_config uart_cfg_##idx = {	              \
+	.base = (struct lpc11u6x_uartx_regs *)                                \
+	DT_REG_ADDR(DT_NODELABEL(uart##idx)),			              \
+	.clock_drv_name =						      \
+	DT_LABEL(DT_PHANDLE(DT_NODELABEL(uart##idx), clocks)),		      \
+	.rx_pinmux_drv_name =                                                 \
+	DT_LABEL(DT_PHANDLE_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs, rxd)),  \
+	.tx_pinmux_drv_name =                                                 \
+	DT_LABEL(DT_PHANDLE_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs, txd)),  \
+	.clkid = DT_PHA_BY_IDX(DT_NODELABEL(uart##idx), clocks, 0, clkid),    \
+	.rx_pin = DT_PHA_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs, rxd, pin), \
+	.rx_func = DT_PHA_BY_NAME(DT_NODELABEL(uart##idx),		      \
+				  pinmuxs, rxd, function),		      \
+	.rx_func = DT_PHA_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs,	      \
+				  rxd, function),			      \
+	.tx_pin = DT_PHA_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs, txd, pin), \
+	.tx_func = DT_PHA_BY_NAME(DT_NODELABEL(uart##idx), pinmuxs,	      \
+				  txd, function),			      \
+	.baudrate = DT_PROP(DT_NODELABEL(uart##idx), current_speed),	      \
+};									      \
+									      \
+static struct lpc11u6x_uartx_data uart_data_##idx;                            \
+									      \
+DEVICE_AND_API_INIT(lpc11u6x_uartx_##idx, DT_LABEL(DT_NODELABEL(uart##idx)),  \
+		    &lpc11u6x_uartx_init,				      \
+		    &uart_data_##idx, &uart_cfg_##idx,			      \
+		    PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS,	      \
+		    &uartx_api)
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay)
+LPC11U6X_UARTX_INIT(1);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart1), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart2), okay)
+LPC11U6X_UARTX_INIT(2);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart2), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart3), okay)
+LPC11U6X_UARTX_INIT(3);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart3), okay) */
+
+#if DT_NODE_HAS_STATUS(DT_NODELABEL(uart4), okay)
+LPC11U6X_UARTX_INIT(4);
+#endif /* DT_NODE_HAS_STATUS(DT_NODELABEL(uart4), okay) */
```

Comment:
```suggestion
DT_INST_FOREACH_STATUS_OKAY(LPC11U6X_UARTX_INIT)
```

Reply:
I'm not sure that would work in my case. On the LPC11U6X MCUs, UART0 has a different interface from the other uart devices, using DT_INST_FOREACH_STATUS_OKAY on LPC11U6X_UARTX_INIT would imply, among other things, initializing all uart devices with the same struct uart_driver_api pointer.
Furthermore, if I understand correctly, there is no guarantee that the instance number Y of a compatible will correspond to the uart Y hardware of the LPC11U6X. In this driver, I need to be sure that the uart device number Y correspond to the actual uart Y of the MCU in order to initialize the shared irq correctly.

Reply:
@henrikbrixandersen Is there a deterministic way to guarantee that instance number X will correspond to the real uartX of the MCU or should we just keep it like this ?

Reply:
No, there are no guarantees in that regard. Perhaps @galak has input on how to go about this?

Reply:
@galak Do you have any insight on this ?